sympy/matrices/random.py

```diff
@@ -0,0 +1,2343 @@
+    2
+
+    Parameters
+    ==========
+    dim : integer
+        dimension of matrix
+    rank : integer (optional with default **dim**-1)
+        rank of matrix
+    scalars : tuple or list of symbols (optional)
+        default values for random choosen scalars of
+        elementary matrices to build the matrix
+    units : tuple or list of symbols (optional)
+        default values for random choosen scalar of
+        elementary matrices to build the matrix
+        which are assumed to have a multiplicative inverse
+    length : integer (optional with default 2 * **dim**)
+        number of invertible matrices to build the resulting matrix.
+
+    See Also
+    ========
+    invertible
+    triangular
+
+    """
+    rank = dim - 1 if rank is None else rank
+    return square(dim, rank, scalars, units, length)
+
+
+# === conjugate matrices, i.e. defined by similarity to a normal from ==
+
+
+def idempotent(dim,
+               rank=None,
+               scalars=_elementary_scalars,
+               units=_elementary_units,
+               length=None):
+    r"""an idempotent square matrix of a given rank s.th. $A*A=A$
+
+    Explanation
+    ===========
+    An idempotent matrix is a matrix $\mathbf{A}$
+    such that $\mathbf{A}^2 = \mathbf{A}$.
+    It is constructed as product
+
+    .. math::
+
+        \mathbf{S} \cdot \mathbf{P} \cdot \mathbf{S}^{-1}
+
+    of an :func:`invertible` matrices $\mathbf{S}$
+    and a :func:`projection` matrix $\mathbf{P}$ of given rank .
+
+    Examples
+    ========
+    >>> from sympy.core.random import seed
+    >>> from sympy.matrices.random import idempotent
+    >>> seed(1)
+
+    >>> A = idempotent(3, 2)
+    >>> A
+    Matrix([
+    [1, 0, 0],
+    [0, 1, 1],
+    [0, 0, 0]])
+    >>> A.rank()
+    2
+    >>> A*A == A
+    True
+
+    Parameters
+    ==========
+    dim : integer
+        dimension of matrix
+    rank : integer (optional with default **dim**)
+        see :func:`projection`
+    scalars : tuple or list of symbols (optional)
+        see :func:`invertible`
+    units : tuple or list of symbols (optional)
+        see :func:`invertible`
+    length : integer (optional with default 2 * **dim**)
+        see :func:`invertible`
+
+    See Also
+    ========
+    elementary
+    projection
+    invertible
+    nilpotent
+
+    References
+    ==========
+    .. [1] https://en.wikipedia.org/wiki/Idempotence#Idempotent_functions
+
+    """
+    rank = dim - 1 if rank is None else rank
+    normal_form = projection(dim, (0, rank))
+    s = invertible(dim, scalars, units, length)
+    return s.inv() * normal_form * s
+
+
+def nilpotent(dim,
+              rank=None,
+              scalars=_elementary_scalars,
+              units=_elementary_units,
+              length=None):
+    r"""a nilpotent matrix of dinemsion n x n.
+
+    Explanation
+    ===========
+    A *nilpotent* matrix is a matrix $\mathbf{A}$ such that there is
+    an integer $n$ with $\mathbf{A}^n = 0$.
```

Comment:
It is perhaps confusing to use `n` in two different ways here.

Reply:
It seems odd that this function allows specifying rank when I would have expected it to allow specifying the `k` such that `A^k = 0` and `A^{k-1} != 0` instead.